// grammarkdown/lib/stringwriter.js
var os_1 = require('os');
var StringWriter = (function () {
    function StringWriter() {
        this._text = "";
        this._depth = 0;
        this._indents = ["", "    "];
        this._newLine = false;
    }
    Object.defineProperty(StringWriter.prototype, "size", {
        get: function () {
            return this._text.length;
        },
        enumerable: true,
        configurable: true
    });
    StringWriter.prototype.indent = function () {
        this._depth++;
    };
    StringWriter.prototype.dedent = function () {
        this._depth--;
    };
    StringWriter.prototype.write = function (text) {
        if (text) {
            this.flushNewLine();
            this._text += text;
        }
    };
    StringWriter.prototype.writeln = function (text) {
        this.write(text);
        this._newLine = true;
    };
    StringWriter.prototype.toString = function () {
        return this._text;
    };
    StringWriter.prototype.flushNewLine = function () {
        if (this._newLine) {
            var indent = this._indents[this._depth];
            if (!indent && this._depth > 0) {
                indent = this._indents[this._depth] = this._indents[this._depth - 1] + this._indents[1];
            }
            this._newLine = false;
            this._text += os_1.EOL + indent;
        }
    };
    return StringWriter;
})();
exports.StringWriter = StringWriter;
