// grammarkdown/lib/tokens.js
/*!
 *  Copyright 2015 Ron Buckton (rbuckton@chronicles.org)
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
var core_1 = require("./core");
(function (SyntaxKind) {
    SyntaxKind[SyntaxKind["Unknown"] = 0] = "Unknown";
    SyntaxKind[SyntaxKind["EndOfFileToken"] = 1] = "EndOfFileToken";
    // Significant Whitespace
    SyntaxKind[SyntaxKind["LineTerminatorToken"] = 2] = "LineTerminatorToken";
    SyntaxKind[SyntaxKind["IndentToken"] = 3] = "IndentToken";
    SyntaxKind[SyntaxKind["DedentToken"] = 4] = "DedentToken";
    // Literals
    SyntaxKind[SyntaxKind["StringLiteral"] = 5] = "StringLiteral";
    SyntaxKind[SyntaxKind["Terminal"] = 6] = "Terminal";
    SyntaxKind[SyntaxKind["UnicodeCharacterLiteral"] = 7] = "UnicodeCharacterLiteral";
    SyntaxKind[SyntaxKind["Prose"] = 8] = "Prose";
    // Punctuation
    SyntaxKind[SyntaxKind["AtToken"] = 9] = "AtToken";
    SyntaxKind[SyntaxKind["OpenBraceToken"] = 10] = "OpenBraceToken";
    SyntaxKind[SyntaxKind["CloseBraceToken"] = 11] = "CloseBraceToken";
    SyntaxKind[SyntaxKind["OpenBracketToken"] = 12] = "OpenBracketToken";
    SyntaxKind[SyntaxKind["CloseBracketToken"] = 13] = "CloseBracketToken";
    SyntaxKind[SyntaxKind["OpenParenToken"] = 14] = "OpenParenToken";
    SyntaxKind[SyntaxKind["CloseParenToken"] = 15] = "CloseParenToken";
    SyntaxKind[SyntaxKind["ColonToken"] = 16] = "ColonToken";
    SyntaxKind[SyntaxKind["ColonColonToken"] = 17] = "ColonColonToken";
    SyntaxKind[SyntaxKind["ColonColonColonToken"] = 18] = "ColonColonColonToken";
    SyntaxKind[SyntaxKind["CommaToken"] = 19] = "CommaToken";
    SyntaxKind[SyntaxKind["PlusToken"] = 20] = "PlusToken";
    SyntaxKind[SyntaxKind["TildeToken"] = 21] = "TildeToken";
    SyntaxKind[SyntaxKind["QuestionToken"] = 22] = "QuestionToken";
    SyntaxKind[SyntaxKind["EqualsToken"] = 23] = "EqualsToken";
    SyntaxKind[SyntaxKind["EqualsEqualsToken"] = 24] = "EqualsEqualsToken";
    SyntaxKind[SyntaxKind["ExclamationEqualsToken"] = 25] = "ExclamationEqualsToken";
    SyntaxKind[SyntaxKind["LessThanExclamationToken"] = 26] = "LessThanExclamationToken";
    SyntaxKind[SyntaxKind["LessThanMinusToken"] = 27] = "LessThanMinusToken";
    // Identifiers
    SyntaxKind[SyntaxKind["Identifier"] = 28] = "Identifier";
    // Keywords
    SyntaxKind[SyntaxKind["ButKeyword"] = 29] = "ButKeyword";
    SyntaxKind[SyntaxKind["EmptyKeyword"] = 30] = "EmptyKeyword";
    SyntaxKind[SyntaxKind["GoalKeyword"] = 31] = "GoalKeyword";
    SyntaxKind[SyntaxKind["HereKeyword"] = 32] = "HereKeyword";
    SyntaxKind[SyntaxKind["ImportKeyword"] = 33] = "ImportKeyword";
    SyntaxKind[SyntaxKind["LexicalKeyword"] = 34] = "LexicalKeyword";
    SyntaxKind[SyntaxKind["LookaheadKeyword"] = 35] = "LookaheadKeyword";
    SyntaxKind[SyntaxKind["NoKeyword"] = 36] = "NoKeyword";
    SyntaxKind[SyntaxKind["NotKeyword"] = 37] = "NotKeyword";
    SyntaxKind[SyntaxKind["OfKeyword"] = 38] = "OfKeyword";
    SyntaxKind[SyntaxKind["OneKeyword"] = 39] = "OneKeyword";
    SyntaxKind[SyntaxKind["OrKeyword"] = 40] = "OrKeyword";
    SyntaxKind[SyntaxKind["ThroughKeyword"] = 41] = "ThroughKeyword";
    // Nodes
    SyntaxKind[SyntaxKind["Parameter"] = 42] = "Parameter";
    SyntaxKind[SyntaxKind["ParameterList"] = 43] = "ParameterList";
    SyntaxKind[SyntaxKind["Argument"] = 44] = "Argument";
    SyntaxKind[SyntaxKind["ArgumentList"] = 45] = "ArgumentList";
    SyntaxKind[SyntaxKind["LinkReference"] = 46] = "LinkReference";
    SyntaxKind[SyntaxKind["Import"] = 47] = "Import";
    SyntaxKind[SyntaxKind["Production"] = 48] = "Production";
    SyntaxKind[SyntaxKind["OneOfList"] = 49] = "OneOfList";
    SyntaxKind[SyntaxKind["RightHandSideList"] = 50] = "RightHandSideList";
    SyntaxKind[SyntaxKind["RightHandSide"] = 51] = "RightHandSide";
    SyntaxKind[SyntaxKind["SymbolSpan"] = 52] = "SymbolSpan";
    // Symbols
    SyntaxKind[SyntaxKind["ButNotSymbol"] = 53] = "ButNotSymbol";
    SyntaxKind[SyntaxKind["UnicodeCharacterRange"] = 54] = "UnicodeCharacterRange";
    SyntaxKind[SyntaxKind["OneOfSymbol"] = 55] = "OneOfSymbol";
    SyntaxKind[SyntaxKind["Nonterminal"] = 56] = "Nonterminal";
    SyntaxKind[SyntaxKind["TerminalList"] = 57] = "TerminalList";
    SyntaxKind[SyntaxKind["SymbolSet"] = 58] = "SymbolSet";
    // Zero-width Assertions
    SyntaxKind[SyntaxKind["EmptyAssertion"] = 59] = "EmptyAssertion";
    SyntaxKind[SyntaxKind["LookaheadAssertion"] = 60] = "LookaheadAssertion";
    SyntaxKind[SyntaxKind["LexicalGoalAssertion"] = 61] = "LexicalGoalAssertion";
    SyntaxKind[SyntaxKind["NoSymbolHereAssertion"] = 62] = "NoSymbolHereAssertion";
    SyntaxKind[SyntaxKind["ParameterValueAssertion"] = 63] = "ParameterValueAssertion";
    // error nodes
    SyntaxKind[SyntaxKind["InvalidSymbol"] = 64] = "InvalidSymbol";
    SyntaxKind[SyntaxKind["InvalidAssertion"] = 65] = "InvalidAssertion";
    SyntaxKind[SyntaxKind["InvalidSourceElement"] = 66] = "InvalidSourceElement";
    // top nodes
    SyntaxKind[SyntaxKind["SourceFile"] = 67] = "SourceFile";
    SyntaxKind[SyntaxKind["FirstKeyword"] = 29] = "FirstKeyword";
    SyntaxKind[SyntaxKind["LastKeyword"] = 41] = "LastKeyword";
    SyntaxKind[SyntaxKind["FirstPunctuation"] = 9] = "FirstPunctuation";
    SyntaxKind[SyntaxKind["LastPunctuation"] = 27] = "LastPunctuation";
})(exports.SyntaxKind || (exports.SyntaxKind = {}));
var SyntaxKind = exports.SyntaxKind;
var textToToken = new core_1.Dict({
    "but": SyntaxKind.ButKeyword,
    "empty": SyntaxKind.EmptyKeyword,
    "goal": SyntaxKind.GoalKeyword,
    "here": SyntaxKind.HereKeyword,
    "import": SyntaxKind.ImportKeyword,
    "lexical": SyntaxKind.LexicalKeyword,
    "lookahead": SyntaxKind.LookaheadKeyword,
    "no": SyntaxKind.NoKeyword,
    "not": SyntaxKind.NotKeyword,
    "of": SyntaxKind.OfKeyword,
    "one": SyntaxKind.OneKeyword,
    "or": SyntaxKind.OrKeyword,
    "through": SyntaxKind.ThroughKeyword,
    "@": SyntaxKind.AtToken,
    ":": SyntaxKind.ColonToken,
    "::": SyntaxKind.ColonColonToken,
    ":::": SyntaxKind.ColonColonColonToken,
    "{": SyntaxKind.OpenBraceToken,
    "}": SyntaxKind.CloseBraceToken,
    "(": SyntaxKind.OpenParenToken,
    ")": SyntaxKind.CloseParenToken,
    "[": SyntaxKind.OpenBracketToken,
    "]": SyntaxKind.CloseBracketToken,
    ",": SyntaxKind.CommaToken,
    "+": SyntaxKind.PlusToken,
    "~": SyntaxKind.TildeToken,
    "?": SyntaxKind.QuestionToken,
    "=": SyntaxKind.EqualsToken,
    "==": SyntaxKind.EqualsEqualsToken,
    "!=": SyntaxKind.ExclamationEqualsToken,
    "<!": SyntaxKind.LessThanExclamationToken,
    "<-": SyntaxKind.LessThanMinusToken,
});
var tokenToText = core_1.Dict.invert(textToToken);
function stringToToken(text) {
    return core_1.Dict.get(textToToken, text);
}
exports.stringToToken = stringToToken;
function tokenToString(kind, quoted) {
    if (kind in tokenToText) {
        var text = tokenToText[kind];
        return quoted ? "'" + text + "'" : text;
    }
    switch (kind) {
        case SyntaxKind.LineTerminatorToken:
            return "«line terminator»";
        case SyntaxKind.IndentToken:
            return "«indent»";
        case SyntaxKind.DedentToken:
            return "«dedent»";
        case SyntaxKind.Identifier:
            return "«identifier»";
        case SyntaxKind.Terminal:
            return "«terminal»";
        case SyntaxKind.UnicodeCharacterLiteral:
            return "«unicode character literal»";
        case SyntaxKind.Prose:
            return "«prose»";
        case SyntaxKind.Production:
            return "«production»";
        case SyntaxKind.OneOfList:
            return "«one of»";
        case SyntaxKind.RightHandSide:
            return "«right hand side»";
    }
    return "«" + SyntaxKind[kind] + "»";
}
exports.tokenToString = tokenToString;
